lib/saluki-config/src/refresher.rs

```diff
@@ -0,0 +1,131 @@
+        Ok(config.as_typed()?)
+    }
+    /// Create `RefreshableConfiguration` from `RefresherConfiguration`.
+    pub fn build(&self) -> Result<Arc<RefreshableConfiguration>, GenericError> {
+        let raw_bearer_token = std::fs::read_to_string(&self.auth_token_file_path)?;
+        let refreshable_configuration = Arc::new(RefreshableConfiguration {
+            token: raw_bearer_token,
+            values: Arc::new(ArcSwap::from_pointee(serde_json::Value::Null)),
+            refresh_interval_seconds: self.refresh_interval_seconds,
+        });
+        refreshable_configuration.clone().spawn_refresh_task();
+        Ok(refreshable_configuration)
+    }
+}
+impl RefreshableConfiguration {
+    /// Start a task that queries the datadog-agent config endpoint every 15 seconds.
+    fn spawn_refresh_task(self: Arc<Self>) {
+        tokio::spawn(async move {
+            let client = ClientBuilder::new()
+                .danger_accept_invalid_certs(true) // Allow invalid certificates
+                .build()
+                .expect("failed to create http client");
+            loop {
+                let response = client
+                    .get(DATADOG_AGENT_CONFIG_ENDPOINT)
+                    .header("Content-Type", "application/json")
+                    .header("Authorization", format!("Bearer {}", self.token))
+                    .header("DD-Agent-Version", "0.1.0")
+                    .header("User-Agent", "agent-data-plane/0.1.0")
+                    .send()
+                    .await;
+                match response {
+                    Ok(response) => {
+                        let config_response: Value = response
+                            .json()
+                            .await
+                            .expect("failed to deserialize configuration into json");
+                        self.values.store(Arc::new(config_response));
+                        debug!("Retrieved configuration from datadog-agent.");
+                    }
+                    Err(e) => {
+                        error!("Error retrieving configuration from datadog-agent: {}", e);
+                    }
+                }
+
+                sleep(Duration::from_secs(self.refresh_interval_seconds)).await;
+            }
+        });
+    }
+
+    /// Gets a configuration value by key.
+    ///
+    ///
+    /// ## Errors
+    ///
+    /// If the key does not exist in the configuration, or if the value could not be deserialized into `T`, an error
+    /// variant will be returned.
+    pub fn get_typed<T>(&self, key: &str) -> Result<T, ConfigurationError>
+    where
+        T: DeserializeOwned,
+    {
+        let values = self.values.load();
+        match values.get(key) {
+            Some(value) => {
+                // Attempt to deserialize the value to type T
+                serde_json::from_value(value.clone()).map_err(|_| ConfigurationError::MissingField {
```

Comment:
The field isn't really missing, it's just not the type we expected. We should be using `ConfigurationError::InvalidFieldType`.

That error variant does want to show the _actual_ type of the field, though, and there's the issue of getting the name of `T`, so you could do something like this:

```rust
ConfigurationError::InvalidFieldType {
  field: key.to_string(),
  expected_ty: std::any::type_name::<T>().to_string(),
  actual_ty: serde_json_value_type_name(value),
}
```

That requires a helper because `serde_json::Value` doesn't have an inherent method to get a stringified name of the variant, so you could add a free function further down like:

```rust
fn serde_json_value_type_name(value: &Value) -> &'static str {
    match value {
        Value::Null => "null",
        Value::Bool(_) => "bool",
        Value::Number(_) => "number",
        Value::String(_) => "string",
        Value::Array(_) => "array",
        Value::Object(_) => "object",
    }
}
```

It won't be perfect because it's not going to help differentiate between arrays of different types, or objects with the wrong structure, and so on... but it should be a good first pass.